### ISSUE TITLE
Bind Optimize API

### ISSUE THREAD

**Issue by jklmnn** [Componolit/AZ3#8]

**Comment by jklmnn** [comment]
The optimization works so far. However trying to optimize a value that would result in an indefinite value makes check never return. E.g. maximizing `A` for `A > 50` will block when checking. A reproducer in C yields the same result:
```C
#include <z3.h>
#include <stdio.h>

int main()
{
    Z3_context context = Z3_mk_context(Z3_mk_config());
    Z3_optimize opt = Z3_mk_optimize(context);
    Z3_ast a = Z3_mk_const(context, Z3_mk_string_symbol(context, "A"), Z3_mk_int_sort(context));
    Z3_ast a42 = Z3_mk_int64(context, 42, Z3_mk_int_sort(context));
    Z3_ast term = Z3_mk_ge(context, a, a42);
    Z3_optimize_assert(context, opt, term);
    Z3_optimize_maximize(context, opt, a);
    printf("check: %d\n", Z3_optimize_check(context, opt));
    Z3_optimize_get_upper(context, opt, 0);
    return 0;
}
```

**Comment by senier** [comment]
When adding a timeout, `Z3_optimize_check` returns 0:

```c
Z3_global_param_set("timeout", "1000" /* in ms */);
```

**Comment by jklmnn** [comment]
Maybe we should add this to the config of our default context instead of globally. This can be done with [`Z3_set_param_value`](https://z3prover.github.io/api/html/group__capi.html#ga89b30e148359a2d97eb0c3e385b2c1c0).

**Comment by senier** [comment]
> Maybe we should add this to the config of our default context instead of globally. This can be done with [`Z3_set_param_value`](https://z3prover.github.io/api/html/group__capi.html#ga89b30e148359a2d97eb0c3e385b2c1c0).

I tried this. Contrary to the documentation my version of Z3 emitted a warning that the parameter "timeout" is unknown.

**Comment by jklmnn** [comment]
The `timeout` parameter only seems to be available for solvers, the correct way to set it for the optimizer seems to be
```C
Z3_config config = Z3_mk_config();
Z3_context context = Z3_mk_context(config);
Z3_params opt_params = Z3_mk_params(context);
Z3_params_set_uint(context, opt_params, Z3_mk_string_symbol(context, "timeout"), 1000);
Z3_optimize opt = Z3_mk_optimize(context);
Z3_optimize_set_params(context, opt, opt_params);
```

However the call to `Z3_optimize_set_params` results in a segmentation fault. I opened an issue in Z3: https://github.com/Z3Prover/z3/issues/4885.

**Comment by jklmnn** [comment]
Fixed in #11.
